zeroconf/__init__.py

```diff
@@ -1073,44 +1082,96 @@ def write_record(self, record: DNSRecord, now: float) -> int:
+        if self.size > len_limit:
             while len(self.data) > start_data_length:
                 self.data.pop()
             self.size = start_size
-            self.state = self.State.finished
-            return 1
-        return 0
+            return False
+        return True
 
     def packet(self) -> bytes:
-        """Returns a string containing the packet's bytes
+        """Returns a string containing the first packet's bytes.
+
+        Generally, you want to use packets() in case the response
+        does not fit in a single packet, but this exists for
+        backward compatibility."""
+        p = self.packets()
+        if len(p) > 0:
+            if len(p[0]) > _MAX_MSG_ABSOLUTE:
+                self.log_warning_once("Created over-sized packet (%d bytes) %r", len(p[0]), p[0])
+            return p[0]
+        else:
+            return b''
+
+    def packets(self) -> List[bytes]:
+        """Returns a list of strings containing the packets' bytes
 
         No further parts should be added to the packet once this
         is done."""
 
-        overrun_answers, overrun_authorities, overrun_additionals = 0, 0, 0
-
-        if self.state != self.State.finished:
+        if self.state == self.State.finished:
+            return self.packets_data
+
+        answer_offset = 0
+        authority_offset = 0
+        additional_offset = 0
+
+        # we have to at least write out the question
+        first_time = True
+
+        while (
+            first_time
+            or answer_offset < len(self.answers)
+            or authority_offset < len(self.authorities)
+            or additional_offset < len(self.additionals)
+        ):
+            first_time = False
+            log.debug("offsets = %d, %d, %d", answer_offset, authority_offset, additional_offset)
+            log.debug("lengths = %d, %d, %d", len(self.answers), len(self.authorities), len(self.additionals))
+
+            additionals_written = 0
+            authorities_written = 0
+            answers_written = 0
+            questions_written = 0
             for question in self.questions:
                 self.write_question(question)
-            for answer, time_ in self.answers:
-                overrun_answers += self.write_record(answer, time_)
-            for authority in self.authorities:
-                overrun_authorities += self.write_record(authority, 0)
-            for additional in self.additionals:
-                overrun_additionals += self.write_record(additional, 0)
-            self.state = self.State.finished
-
-            self.insert_short(0, len(self.additionals) - overrun_additionals)
-            self.insert_short(0, len(self.authorities) - overrun_authorities)
-            self.insert_short(0, len(self.answers) - overrun_answers)
-            self.insert_short(0, len(self.questions))
+                questions_written += 1
+            allow_long = True  # at most one answer is allowed to be a long packet
```

Comment:
Is it at most one answer per packet or per group of packets?

Reply:
Added docstring -- multiple answers per packet are fine.  It's just the case of having so many answers that it overflows the smaller 1470ish MTU of ethernet -- in that scenario, the application layer MUST send separate application-level responses rather than rely on IP fragmentation to break up a bigger packet.  The only exception is when a single answer does not fit inside the smaller MTU in which case it's allowed to be bigger.  I suspect that scenario will still crash ChromeCast Audios, but *that* situation is exceedingly rare, whereas many responses that exceed 1470 bytes happens within a large but not crazy set of devices.